Expand Operation Concepts and Create Administration Concepts doc (#519)
minio/docs: source/administration/concepts.rst

MinIO buckets provide the same functionality as AWS S3 buckets. 

For example, consider an application that hosts a web blog. 
The application needs to store a variety of blobs, including rich multimedia like videos and images. 
The structure of objects on the MinIO server might look similar to the following:

.. code-block:: text

   / #root
   /images/
      2020-01-02-MinIO-Diagram.png
      2020-01-03-MinIO-Advanced-Deployment.png
      MinIO-Logo.png
   /videos/
      2020-01-04-MinIO-Interview.mp4
   /articles/
      /john.doe/
         2020-01-02-MinIO-Object-Storage.md
         2020-01-02-MinIO-Object-Storage-comments.json
      /jane.doe/
         2020-01-03-MinIO-Advanced-Deployment.png
         2020-01-02-MinIO-Advanced-Deployment-comments.json
         2020-01-04-MinIO-Interview.md

MinIO supports multiple levels of nested directories and objects to support even the most dynamic object storage workloads.

Encryption
----------

Encryption-at-Rest
   A method of data security that stores objects on a disk in an encoded state.
   To read the data, the server must first decrypt the data using a decryption key.

> Encryption-in-Transit

Author: Added.

> :ref:`Identify and access management (IAM) <minio-authentication-and-identity-management>` is therefore a critical component of a MinIO configuration.

Author: Need to correct this to `Identity and access management ...`

Specifically, clients must present a valid access key and secret key to access any S3 or MinIO administrative API, such as ``PUT``, ``GET``, and ``DELETE`` operations. 

MinIO then checks that authenticated users or clients have *authorization* to perform actions or use resources on the deployment. 
MinIO uses :ref:`Policy-Based Access Control (PBAC) <minio-policy>`, where each policy describes one or more rules that outline the permissions of a user or group of users. 

> MinIO supports S3-specific:ref:`actions <minio-policy-actions>` and :ref:`conditions <minio-policy-conditions>` when creating policies. 

Author: First :ref: link needs a space before it.

   TLS provides the standards used by internet communications to support encryption, authentication, and data integrity through a TLS handshake.
   The handshake confirms the identify of the requesting party and whether the party should have access to the requested data.

   After confirming identity and access, TLS provides the cipher to encrypt and then decrypt the transfer of information from the server to the requesting client.

   TLS succeeds and replaces the previously used Secure Socket Layer (SSL)

   MinIO supports versions 1.2 and 1.3 of the TLS standard.

Object Management
-----------------

:ref:`Information Lifecycle Management (ILM) <minio-lifecycle-management>`
   Information Lifecycle Management establishes rules to determine where to retain an object and how long to retain it.

>    For exaple, some objects may need to be retained for legal reasons for a specific period of time before anyone can delete them.

Author: Nice catch. Thanks.

>    Other objects should be deleted after three days.

Author: It's an alternative example. Added "Or in other circumstances" to better link it as part of a set of three examples.

>    Objects that are accessed frequently should go in warm storage while those needed only for archival purposes and rarely accessed can move to cold storage.

Author: Good idea! I added a section for "Storage Temperatures" with some explanatory verbiage.

> MinIO supports methods that encode objects while on disk (encryption-at-rest) and that encode objects during transition from one location to another (encryption-in-transit, or "in flight").

Author: Cut the second "that encode objects" to simplify the sentence.

   ILM is the set of rules that determine when to move objects between storage classes and how long to retain objects or prevent their deletion.

Locking
   MinIO uses a method of object retention called Object Locking.
   This Write Once Read Many (WORM) mechanism protects an object's immutability, preventing any changes to an object while in a locked state.

   Locks can be based on a specified period of time for unprivileged users (governance mode) or for all users (compliance mode).
   Administrations can use a legal hold to retain objects indefinitely, which prevents all users from deleting or modifying the locked object.

Prefix

>    Prefixes are the beginning part of an object name.

Author: Hmm, not really.
Per Amazon's docs, S3 is a flat structure with no real hierarchy whatsoever.
Folders are simply an idea construed from a common string at the beginning of the object name with a specified delimiter character. The common string (the prefix) is, indeed, part of the object's _key name_.

See https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-folders.html and also https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html.

> In Amazon S3, buckets and objects are the primary resources, and objects are stored in buckets. **Amazon S3 has a flat structure instead of a hierarchy like you would see in a file system.** However, for the sake of organizational simplicity, the Amazon S3 console supports the folder concept as a means of grouping objects. **It does this by using a shared name prefix for objects (that is, objects have names that begin with a common string).** Object names are also referred to as key names.

A "folder" is nothing more than a 0 byte object added for the convenience of the user's mental model that comes from our long history of traditional file systems.

   They use the ``/`` character to indicate separate levels of differentiation from other objects.
   You can use prefixes to add structure and a hierarchy to the objects in a MinIO bucket.

>    MinIO parses the name of an object to identify prefixes, then creates a structure to bucket contents that mimics the folder and subfolder directory structure of traditional file systems.

Author: 👍

   Consider an object in an ``images`` bucket named ``vacation/2019/october/17/mountain-hike.jpg``.
   MinIO treats this as a file ``mountain-hike.jpg`` in a nested series of locations:

   .. code-block::

      vacation/
         2019/
            october/
               17

   However, MinIO buckets do not actually arrange contents in folders.
   You must always refer to an object with its full name, including the prefix.
   For MinIO objects, the prefix "path" is part of the object name.

Replication
   Duplicating the object data from one MinIO deployment location to another MinIO deployment location.
   
   :ref:`Bucket level replication <minio-bucket-replication>` duplicates the objects in one bucket to another.
   These can be separate buckets on the same deployment or across different deployments.

   MinIO supports two methods of bucket replication:

   #. Active-Passive replication performs one-way duplication from one location to another. 
      Changes on the second location do not return to the first location.

>    #. Active-Active replication performs bi-directional, two-way replication between two locations.

Author: Multi-site bucket replication is really just setting up two-way replication between each possible pairing of deployments. So, I feel like this is still accurate.

>       Changes on either location replicate to the other location.

Author: I can call that out.

   :ref:`Site level replication <minio-site-replication-overview>` keeps two or more separate MinIO deployments in sync.
   All buckets and objects on each deployment duplicate to the other sites configured for replication.

Retention
   Some situations require that certain objects be protected from either change or deletion.
   MinIO uses retention tools to establish rules for whether to allow modification or deletion of an object's data.

   MinIO supported retention tools include versioning and locking.

Versioning
   Ordinarily, a write operation to a specific name that already exists results in overwriting the previously existing object.

   MinIO can maintain access to various iterations of a uniquely named object as it changes over time through versioning.
   When enabled, MinIO writes mutated objects to a different version than the original, allowing access to both the original object and the newer, changed object.

   Configurations on the MinIO bucket determine how long to retain older versions of an object.

> WORM

Author: Yes, it is redundant. I've moved this section to the separate Glossary page I'm adding.

Blobs can be images, audio files, spreadsheets, or even binary executable code. 
Object Storage platforms like MinIO provide dedicated tools and capabilities for storing, retrieving, and searching for blobs. 

.. _buckets:

MinIO Object Storage uses :ref:`buckets <buckets>` to organize objects. 
A bucket is similar to a folder or directory in a filesystem, where each bucket can hold an arbitrary number of objects. 
MinIO buckets provide the same functionality as AWS S3 buckets. 

For example, consider an application that hosts a web blog. 
The application needs to store a variety of blobs, including rich multimedia like videos and images. 
The structure of objects on the MinIO server might look similar to the following:

.. code-block:: text

   / #root
   /images/
      2022-01-02-MinIO-Diagram.png
      2022-01-03-MinIO-Advanced-Deployment.png
      MinIO-Logo.png
   /videos/
      2022-01-04-MinIO-Interview.mp4
   /articles/
      /john.doe/
         2022-01-02-MinIO-Object-Storage.md
         2022-01-02-MinIO-Object-Storage-comments.json
      /jane.doe/
         2022-01-03-MinIO-Advanced-Deployment.png
         2022-01-02-MinIO-Advanced-Deployment-comments.json

>          2022-01-04-MinIO-Interview.md

Reviewer: Suggest dropping these examples and using them when we talk about prefixing, as it makes way more sense there.

Author: I like this idea. I've replaced the example in the prefix section with this content.

MinIO supports **Transport Layer Security** (TLS) versions 1.2 and 1.3 encrypting objects.
TLS replaces the previously used Secure Socket Layer (SSL) method that has since been deprecated.
The TLS standard, maintained by the Internet Engineering Task Force (IETF), provides the standards used by internet communications to support encryption, authentication, and data integrity.

The process of authenticating a user and verifying access to objects is known as the `TLS Handshake`.
Once authenticated, TLS provides the cipher to encypt and then decrypt the transfer of information from the server to the requesting client.

MinIO supports several methods of :ref:`Server-Side Encryption <minio-encryption-overview>`.

Can I organize objects in a folder structure within buckets?
------------------------------------------------------------

MinIO is a flat object store, meaning there is no structure to a deployment other than the bucket(s).
However, MinIO does allow for using a prefixing scheme with a ``/`` delimitor that can mimic a folder structure from traditional file systems.

Prefixing involves prepending the name of an object with a fixed string.

> When you use the same prefix string on multiple objects, MinIO identifies those as similar or grouped objects.

Reviewer: I think calling MinIO a 'flat object store', while technically true, may be misleading. I _think_ the distinction is more that:

- MinIO supports any arbitrary hierarchy of objects in a bucket
- You do not create "subfolders" - you use prefixes as part of the object name
- MinIO infers from that full path what the "subfolders" are
- Multiple objects with similar prefix paths "look" like folder/subfolder relationships from the POSIX world
- All of this is automatic.

Reviewer: I'm not able to be more prescriptive *yet*. The prefixing stuff is vital, and hard to explain.

Author: The water is definitely muddied because we use prefixes, but call them folders and subfolders (very much against Amazon there). I reworked this section.

Consider an object in an ``images`` bucket named ``vacation/2019/october/17/mountain-hike.jpg``.
MinIO treats this as an object ``mountain-hike.jpg`` in a nested series of locations:

.. code-block::

   vacation/
      2019/
         october/
            17

> If you had a second object in the bucket named ``vacation/2019/october/17/waterfall.jpg``, you could find both objects by calling on the ``vacation/2019/october/17/`` prefix.

Reviewer: ```suggestion
If you had a second object in the bucket named ``vacation/2019/october/17/waterfall.jpg``, you could find both objects by performing a LIST operation on the ``vacation/2019/october/17/`` prefix.
```

Author: 👍

Remember:

> - MinIO buckets do not actually arrange contents in folders.

Reviewer: Not sure what this section is intended to get across.

Author: cut

- You must always refer to a specifc object with its full name, including the prefix.
- For MinIO objects, the prefix "path" is part of the object name.

> What tools does MinIO provide to manage where objects store?

Reviewer: awk - maybe:

```suggestion
How can I manage objects stored on MinIO?
```

Also I would not group replication and tiering together.

Tiering and Expiration both would fall under "How do I manage objects over time?"

Replication is strictly "How do I backup and restore MinIO?" with a clear and concise message that you don't do it via traditional methods, you use replication and resynchronization.

Author: Moved tiering to the next section.
Added mirroring to this section and retitled it `How can I backup and restore objects on MinIO?`

------------------------------------------------------------

MinIO provides two features to manage where an object stores.
These tools include replication and tiering.

MinIO provides two types of replication to copy an object from one location to another.
You can configure replication at either the :ref:`bucket level <minio-bucket-replication>` or at the :ref:`site level <minio-site-replication-overview>`.

- Bucket level replication can function as either one-way, active-passive replication (such as for archival purposes) or as two-way, active-active replication to keep two buckets in sync with each other.
- Site level replication functions as two-way, active-active replication to keep multiple data locations (such as different geographic data centers) in sync with one another.

:ref:`Tiering rules <minio-lifecycle-management-tiering>` allow frequently accessed objects to store on hot or warm storage, which is typically more expensive but provides better performance.
Less frequently accessed objects can move to cold storage.
Cold storage often exchanges slower performance for a cheaper price.

> What tools does MinIO provide to manage object lifecycle?

Reviewer: If we are keeping this, then the previous section can maybe specifically target "How do I use different storage types with MinIO?" because the answer *there* is to use tiering with hot/warm/cold MinIO deployments based on storage type.

Reviewer: Oh - and this section should be ILM only.

Locking/Versioning is more "How does MinIO protect objects from accidental overwrite or deletion?"

Author: Split as suggested.
